server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableAction.java

```diff
@@ -82,6 +90,30 @@ protected void doExecute(Task task, Request request, ActionListener<ActionRespon
         }
     }
 
+    private void failShard(ShardRouting shardRouting, ClusterState clusterState, ActionListener<Object> l, Exception e) {
```

Comment:
Can we parameterize this with the specific type? `ActionListener<TransportResponse.Empty> l`

I think if you change the call site a bit it should be possible:

```
                request.indexShardRoutingTable.unpromotableShards().forEach(shardRouting -> {
                    final DiscoveryNode node = clusterState.nodes().get(shardRouting.currentNodeId());
                    ActionListener<TransportResponse.Empty> acquired = listeners.acquire(ignored -> {});
                    transportService.sendRequest(
                        node,
                        transportUnpromotableAction,
                        request,
                        TransportRequestOptions.EMPTY,
                        new ActionListenerResponseHandler<>(
                            request.failShardOnError()
                                ? acquired.delegateResponse((l, e) -> failShard(shardRouting, clusterState, l, e))
                                : acquired,
                            (in) -> TransportResponse.Empty.INSTANCE,
                            executor
                        )
                    );
                });
```